### PR TITLE
Create new product category to support enrollment codes

### PR DESCRIPTION
**Pull request by vkaracic**
Created a new Client model to use for the enrollment codes. Added a new product class type via a migration. Added a `courses` field to the catalog model because a catalog will have one or more courses and a course will have one or more catalogs. Accompanying tests are included. 

https://openedx.atlassian.net/browse/SOL-1326

@mjfrey @clintonb please take a look.

### CODE REVIEW

File: ecommerce/extensions/catalogue/models.py [hunk 1]
```diff
@@ -20,6 +20,7 @@ class Catalog(models.Model):
     name = models.CharField(max_length=255)
     partner = models.ForeignKey('partner.Partner', related_name='catalogs')
     stock_records = models.ManyToManyField('partner.StockRecord', blank=True, related_name='catalogs')
+    courses = models.ManyToManyField('courses.Course', blank=True)
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/models.py]: Why is this necessary? You can get to courses through the stock records and related products. A catalog is the general concept of a catalog. It includes more than just courses, so should share no direct links to courses.

File: ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('catalogue', '0001_initial'),
+        ('catalogue', '0011_auto_20151019_0639')
+    ]
+    operations = [
+        migrations.RunPython(create_product_class)
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py]: Include a reverse operation.

File: ecommerce/core/models.py [hunk 3]
```diff
@@ -59,3 +59,8 @@ class Meta(object):
 
     def get_full_name(self):
         return self.full_name or super(User, self).get_full_name()
+
+
+class Client(User):
+    """Custom user model for clients."""
```
Comment (clintonb) [inline ecommerce/core/models.py]: This could use a better docstring. Explain why the model exists.

File: ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py (hunk 2 above)
Comment (mjfrey) [inline ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py]: Do we need reverse for all migrations?  I do not see one for the Seat migration.

Comment (clintonb) [inline ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py]: Yes, all migrations should have reverse migrations. Feel free to add one for the seat migration. That was created before I was on the team.

**Comment by vkaracic** [comment]
@clintonb I have reverted the commit that added a new field to Catalog, addressed your other comments and changed the tests accordingly. Can you please take a look now?

**Comment by mjfrey** [comment]
LGTM now.
:+1:

File: .travis.yml [hunk 4]
```diff
@@ -36,6 +36,6 @@ script:
     - make validate_js
 branches:
     only:
-        - master
+        - ecodes/master
```
Comment (clintonb) [inline .travis.yml]: Remove.

File: ecommerce/core/models.py [hunk 5]
```diff
@@ -59,3 +59,8 @@ class Meta(object):
 
     def get_full_name(self):
         return self.full_name or super(User, self).get_full_name()
+
+
+class Client(User):
+    """The model for the customer that purchases enrollment codes."""
```
Comment (clintonb) [inline ecommerce/core/models.py]: `Business clients/purchasers.`

This model doesn't care about enrollment codes, so there should be no mention of enrollment codes in its comment.

File: ecommerce/extensions/catalogue/migrations/0002_auto_20150223_1052.py [hunk 6]
```diff
@@ -60,6 +60,19 @@ def create_catalog(apps, schema_editor):
...
 
 
+def remove_catalog(apps, schema_editor):
+    """ Reverse function. """
+    Category = apps.get_model("catalogue", "Category")
+    ProductAttribute = apps.get_model("catalogue", "ProductAttribute")
+    ProductClass = apps.get_model("catalogue", "ProductClass")
+
+    Category.objects.filter(slug='seats').delete()
+    ProductAttribute.objects.filter(code='certificate_type').delete()
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/migrations/0002_auto_20150223_1052.py]: Delete the `ProductClass` and the attributes will be deleted automatically.

File: ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def remove_product_class(apps, schema_editor):
+    """ Reverse function. """
+    Category = apps.get_model("catalogue", "Category")
+    ProductAttribute = apps.get_model("catalogue", "ProductAttribute")
+    ProductClass = apps.get_model("catalogue", "ProductClass")
+
+    Category.objects.filter(slug='enrollment_code').delete()
+    ProductAttribute.objects.filter(code='end_date').delete()
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py]: Same as above. You only need to delete the `ProductClass`.

**Comment by vkaracic** [comment]
@clintonb updated!

File: ecommerce/extensions/catalogue/tests/test_enrollment_codes.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    def test_enrollment_code_product(self):
+        """
+        Test if an enrollment code is properly created
+        and has a course associated with it.
+        """
+        catalog = Catalog.objects.create(
+            partner_id=self.partner.id
+        )
+        course = Course.objects.create(
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/tests/test_enrollment_codes.py]: Consider using `CourseFactory`.

File: ecommerce/extensions/catalogue/tests/test_enrollment_codes.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            id='org/number/run',
+            name='Test course'
+        )
+
+        ec_product_class = ProductClass.objects.get(slug='enrollment_code')
+        enrollment_code_product = Product.objects.create(
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/tests/test_enrollment_codes.py]: Consider using `ProductFactory`.

File: ecommerce/extensions/catalogue/tests/test_enrollment_codes.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import datetime
+from oscar.core.loading import get_model
```
Comment (clintonb) [inline ecommerce/extensions/catalogue/tests/test_enrollment_codes.py:3]: Imports should be grouped: main lib, third-party, local. Each group should be alphabetized.

**Comment by clintonb** [comment]
:+1:

File: ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    Category.objects.create(
+        description="All Enrollment Codes",
+        numchild=2,
+        slug="enrollment_codes",
+        depth=1,
+        path="0002",
+        image="",
+        name="Enrollment Codes"
+    )
+
```
Comment (mjfrey) [inline ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py:91]: I am just thinking we missed an attribute we need at fulfillment time.  The type of Voucher to create. Multi-use, Single-use etc...
@clintonb what do you think?
I suggest adding an attribute called "Type" and having it match the ENUM that Abstract Voucher uses.

Comment (clintonb) [inline ecommerce/extensions/catalogue/migrations/0012_enrollment_code_product_class.py:91]: Agreed.

## 

Clinton Blackburn

**Comment by vkaracic** [comment]
@mjfrey @clintonb I have assigned a new product attribute of type option to the enrollment code product class. Please take a look at the last commit. If you are happy with the code I believe I can merge this.

**Comment by clintonb** [comment]
:+1:

**Comment by mjfrey** [comment]
Great.

:+1:
